### PR TITLE
Add URL parser for scraper endpoint

### PR DESCRIPTION
**Pull request by phhai-hcmut**
Also support tweet URL

### CODE REVIEW

File: app/api/scraper.py [hunk 1]
```diff
@@ -2,23 +2,23 @@
...
 @router.get("/check", response_model=schemas.CheckResponse, name="user:get-data")
 def user_info_check(url: str, scraper: TwitterScraper = Depends()):
-    # Validate input url
-    if url[:20] == "https://twitter.com/":
-        username = url.split("/")[3]
-    elif url[:12] == "twitter.com/":
-        username = url.split("/")[1]
-    elif url != "" and "/" not in url:
-        username = url
-    else:
```
Comment (dirtyg2120) [inline app/api/scraper.py:18]: u forgot to remove same thing as this one in line 31->37, I can't comment on that :V

File: app/api/scraper.py [hunk 2]
```diff
@@ -2,23 +2,23 @@
...
+        twitter_entity, twitter_id = parse_twitter_url(url)
+    except ValueError:
         raise HTTPException(status_code=400, detail="'url' argument is invalid!")
 
-    user_db = scraper.get_user_by_username(username)
+    if twitter_entity == "tweet":
+        tweet = scraper.api_v2.get_tweet(twitter_id, tweet_fields=["author_id"])
+        user_db = scraper.get_user_by_id(tweet.data.author_id)
+    else:
+        user_db = scraper.get_user_by_username(twitter_id)
```
Comment (produdez) [inline app/api/scraper.py:21]: Just realize that in the `tweetUrl` contains the username :D, so we can just get the username and everything is back to normal ✌️. No need to the extra flow `twitter_entity == 'tweet`

File: app/services/url_parser.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+from urllib.parse import urlparse
+
+from fastapi import HTTPException
+
+
+def parse_twitter_url(url: str) -> Tuple[Literal["user", "tweet"], str]:
+    parsed = urlparse(url)
+
+    if parsed.netloc != "twitter.com":
+        raise HTTPException(400, "'url' argument is invalid!")
```
Comment (produdez) [inline app/services/url_parser.py:11]: Wait, does this mean that calling `check` endpoint with `username` is invalid now? That' not good

**Review by produdez — CHANGES_REQUESTED** [review]

File: app/api/scraper.py (hunk 2 above)
Comment (phhai-hcmut) [reply app/api/scraper.py:21]: Why not need extra flow?

Comment (produdez) [reply app/api/scraper.py:21]: Cause if we just get the `username` from the `twitterUrl`, things get simple?

Comment (phhai-hcmut) [reply app/api/scraper.py:21]: There is no `twitterUrl` in this function. Besides, we need to support tweet URL too.

Comment (produdez) [reply app/api/scraper.py:21]: yes by `twitterUrl` i ment tweet URL (sr for confusion). Point being we can get the account username from the tweet URL. It simplifies everything that way!

**Comment by phhai-hcmut** [comment]
@dirtyg2120 @produdez Check again please

File: app/api/scraper.py (hunk 1 above)
Comment (dirtyg2120) [reply app/api/scraper.py]: @phhai-hcmut did u forget this?
on `user_detail_check()`

File: app/api/scraper.py [hunk 4]
```diff
@@ -28,15 +26,13 @@ def user_info_check(url: str, scraper: TwitterScraper = Depends()):
...
-        username = url.split("/")[1]
-    elif url != "" and "/" not in url:
+    if url == "":
+        raise HTTPException(400, "'url' argument is invalid!")
+
+    if "/" not in url:
         username = url
     else:
-        raise HTTPException(status_code=400, detail="'url' argument is invalid!")
+        username = parse_twitter_url(url)
```
Comment (produdez) [inline app/api/scraper.py:35]: Cant we put all the input validation logic in the same module? Seeing half of it outside and half of it inside the `parse_twitter_url` is kinda itchy

Comment (phhai-hcmut) [reply app/api/scraper.py:35]: Because bare username is not URL?

Comment (produdez) [reply app/api/scraper.py:35]: yea but :v, cant we just rename the function to `parse_twitter_username` :v?

**Review by dirtyg2120 — DISMISSED** [review]

**Review by produdez — APPROVED** [review]

**Pull request merged**